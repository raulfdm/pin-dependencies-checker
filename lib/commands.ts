import parseArguments from "yargs-parser";
import { z } from "zod";

const argv = parseArguments(process.argv.slice(2));

const CliConfig = z.object({
	peerDeps: z.boolean().default(false),
	deps: z.boolean().default(true),
	devDeps: z.boolean().default(true),
	optionalDeps: z.boolean().default(false),
});

export type CliConfigType = z.infer<typeof CliConfig>;

export const cliConfig = CliConfig.parse(argv);
