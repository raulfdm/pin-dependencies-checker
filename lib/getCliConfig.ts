import { parseArgs } from "node:util";
import type { RequiredDeep } from "type-fest";

/**
 * Even when we specify the default value, parseArgs will return
 * a type of `boolean | undefined` for the option.
 *
 * Here we just ensure to TS that the type will be a boolean `boolean`.
 */
export type CliConfigType = RequiredDeep<ReturnType<typeof getCliConfig>>;

export function getCliConfig() {
	const { values } = parseArgs({
		args: process.argv.slice(2),
		options: {
			"no-deps": {
				type: "boolean",
				default: false,
			},
			"no-dev-deps": {
				type: "boolean",
				default: false,
			},
			"optional-deps": {
				type: "boolean",
				default: false,
			},
			"peer-deps": {
				type: "boolean",
				default: false,
			},
		},
	});

	return values;
}
