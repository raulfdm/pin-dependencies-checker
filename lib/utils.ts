export function exitWithSuccess(): void {
	process.exit(0);
}

export function exitWithError(): void {
	process.exit(1);
}

// biome-ignore lint/suspicious/noExplicitAny: <explanation>
export function log(message?: any, ...optionalParams: any[]): void {
	console.log(message, ...optionalParams);
}
