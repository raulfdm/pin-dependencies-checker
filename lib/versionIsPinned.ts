const semverPattern =
	/^(\d+)\.(\d+)\.(\d+)(-[\w-]+(\.[\w-]+)*)?(\+[\w-]+(\.[\w-]+)*)?$/;

const containsCommitishPattern = /[a-f0-9]{5,40}/;

const containsSemverPattern =
	/(\d+)\.(\d+)\.(\d+)(-[\w-]+(\.[\w-]+)*)?(\+[\w-]+(\.[\w-]+)*)?/;

function isUrl(version: string): boolean {
	return version.includes("/");
}

export function versionIsPinned(version: string) {
	if (version === "" || version === "latest" || version === "*") {
		return false;
	}

	const firstCharacter = version[0] ?? "";

	if (["^", ">", "<", "~"].includes(firstCharacter)) {
		return false;
	}

	if (version.startsWith("file:")) {
		return true;
	}

	if (version.startsWith("workspace:")) {
		return semverPattern.test(version.substring(10));
	}

	if (isUrl(version)) {
		return (
			containsSemverPattern.test(version) ||
			containsCommitishPattern.test(version)
		);
	}

	return semverPattern.test(version) || isUrl(version);
}
