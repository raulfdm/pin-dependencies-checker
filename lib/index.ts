import { lib } from "./lib";

await lib();
