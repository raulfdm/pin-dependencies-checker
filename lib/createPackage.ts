import type { Package as GetPackage } from "@manypkg/get-packages";
import { getCliConfig } from "./getCliConfig";
import { versionIsPinned } from "./versionIsPinned";

export type Package = ReturnType<typeof createPackage>;

export function createPackage(pkg: GetPackage) {
	const cliConfig = getCliConfig();

	type Dependency = {
		name: string;
		version: string;
	};

	const unpinnedList: Dependency[] = [];

	const allDependencies = new Map<string, string>();

	if (cliConfig["no-deps"] === false && pkg.packageJson.dependencies) {
		for (const [dep, version] of Object.entries(pkg.packageJson.dependencies)) {
			allDependencies.set(dep, version);
		}
	}

	if (cliConfig["no-dev-deps"] === false && pkg.packageJson.devDependencies) {
		for (const [dep, version] of Object.entries(
			pkg.packageJson.devDependencies,
		)) {
			allDependencies.set(dep, version);
		}
	}

	if (cliConfig["peer-deps"] && pkg.packageJson.peerDependencies) {
		for (const [dep, version] of Object.entries(
			pkg.packageJson.peerDependencies,
		)) {
			allDependencies.set(dep, version);
		}
	}

	if (cliConfig["optional-deps"] && pkg.packageJson.optionalDependencies) {
		for (const [dep, version] of Object.entries(
			pkg.packageJson.optionalDependencies,
		)) {
			allDependencies.set(dep, version);
		}
	}

	for (const [dependency, version] of allDependencies.entries()) {
		if (!versionIsPinned(version)) {
			unpinnedList.push({
				name: dependency,
				version,
			});
		}
	}

	return {
		path: `${pkg.dir}/package.json`,
		getUnpinnedDependencies,
		hasUnpinnedDependency,
	};

	function getUnpinnedDependencies(): Dependency[] {
		return [...unpinnedList];
	}

	function hasUnpinnedDependency(): boolean {
		return unpinnedList.length > 0;
	}
}
