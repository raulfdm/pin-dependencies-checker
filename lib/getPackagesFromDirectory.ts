import { getPackages } from "@manypkg/get-packages";
import { createPackage } from "./createPackage";

export async function getPackagesFromDirectory() {
	const { packages, rootPackage } = await getPackages(process.cwd());

	const allPackages = packages.map(createPackage);

	/**
	 * In monorepos, rootPackage is not included in the `packages` list,
	 * so we have to include it manually.
	 */
	if (rootPackage) {
		allPackages.unshift(createPackage(rootPackage));
	}

	return allPackages;
}
