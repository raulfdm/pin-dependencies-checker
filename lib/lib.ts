import type { Package } from "./createPackage";
import { getPackagesFromDirectory } from "./getPackagesFromDirectory";
import { exitWithError, exitWithSuccess, log } from "./utils";

export async function lib() {
	const allPackages = await getPackagesFromDirectory();

	const allUnpinned = allPackages.filter((pkg) => pkg.hasUnpinnedDependency());

	if (allUnpinned.length > 0) {
		printUnpinnedDependencies(allUnpinned);
		exitWithError();
	} else {
		log("It seems all your dependencies are pinned :)");
		exitWithSuccess();
	}
}

function printUnpinnedDependencies(pkgs: Package[]) {
	for (const packagesWithUnpinned of pkgs) {
		log(`Package: ${packagesWithUnpinned.path}`);

		const unpinnedDeps = packagesWithUnpinned.getUnpinnedDependencies();

		for (const prodDeps of unpinnedDeps) {
			log(`â†’ ${prodDeps.name}@${prodDeps.version}`);
		}
	}
}
