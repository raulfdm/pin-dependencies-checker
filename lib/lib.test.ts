import type { Packages as GetPackages } from "@manypkg/get-packages";
import type { CliConfigType } from "./getCliConfig";
import { lib } from "./lib";

const mockGetPackages = vi.fn();
vi.mock("@manypkg/get-packages", () => ({
	getPackages: () => mockGetPackages(),
}));

const mockGetCliConfig = vi.fn();
vi.mock("./getCliConfig", () => ({
	getCliConfig: () => mockGetCliConfig(),
}));

const mockExitWithSuccess = vi.fn();
const mockExitWithError = vi.fn();
const mockLog = vi.fn();
vi.mock("./utils", () => ({
	exitWithSuccess: () => mockExitWithSuccess(),
	exitWithError: () => mockExitWithError(),
	// biome-ignore lint/suspicious/noExplicitAny: I don't care here
	log: (...args: any) => mockLog(...args),
}));

describe("lib", () => {
	beforeEach(() => {
		doMockCommands();
		doMockGetPackages();
	});

	describe("no pinned lib", () => {
		it("exits process with 0", async () => {
			await lib();
			expect(mockExitWithSuccess).toHaveBeenCalled();
		});

		it("prints a nice message", async () => {
			await lib();

			expect(mockLog.mock.calls.flat()).toMatchInlineSnapshot(`
				[
				  "It seems all your dependencies are pinned :)",
				]
			`);
		});
	});

	describe("found pinned lib", () => {
		it("exits process with 1", async () => {
			doMockGetPackages({
				packages: [
					{
						dir: "/path/to/dir",
						relativeDir: "dir",
						packageJson: {
							name: "test",
							version: "1.0.0",
							dependencies: {
								"test-dep": "^1.0.0",
							},
						},
					},
				],
			});

			await lib();
			expect(mockExitWithError).toHaveBeenCalled();
		});

		describe("prod dependencies", () => {
			beforeEach(() => {
				doMockGetPackages({
					packages: [
						{
							dir: "/path/to/dir",
							relativeDir: "dir",
							packageJson: {
								name: "test",
								version: "1.0.0",
								dependencies: {
									"test-dep": "^1.0.0",
								},
							},
						},
					],
				});
			});

			it("prints all pinned dependencies by default", async () => {
				await lib();

				expect(mockLog.mock.calls.flat()).toMatchInlineSnapshot(`
					[
					  "Package: /path/to/dir/package.json",
					  "→ test-dep@^1.0.0",
					]
				`);
			});

			it("does not fail if --deps=false but there are unpinned production dependencies", async () => {
				doMockCommands({
					"no-deps": true,
				});

				await lib();

				expect(mockExitWithSuccess).toHaveBeenCalled();
			});
		});

		describe("dev dependencies", () => {
			beforeEach(() => {
				doMockGetPackages({
					packages: [
						{
							dir: "/path/to/dir",
							relativeDir: "dir",
							packageJson: {
								name: "test",
								version: "1.0.0",
								devDependencies: {
									"test-dev-dep": "^2.0.0",
									"another-test-dev-dep": "^1.0.0",
								},
							},
						},
					],
				});
			});

			it("prints all pinned dependencies by default", async () => {
				await lib();

				expect(mockLog.mock.calls.flat()).toMatchInlineSnapshot(`
					[
					  "Package: /path/to/dir/package.json",
					  "→ test-dev-dep@^2.0.0",
					  "→ another-test-dev-dep@^1.0.0",
					]
				`);
			});

			it("does not print pinned dependencies if --no-dev-deps", async () => {
				doMockCommands({
					"no-dev-deps": true,
				});

				await lib();

				expect(mockExitWithSuccess).toHaveBeenCalled();
			});
		});

		describe("peer dependencies", () => {
			beforeEach(() => {
				doMockGetPackages({
					packages: [
						{
							dir: "/path/to/dir",
							relativeDir: "dir",
							packageJson: {
								name: "test",
								version: "1.0.0",
								peerDependencies: {
									"react-dom": "^18.0.0",
								},
							},
						},
					],
				});
			});

			it("does not print pinned dependencies by default", async () => {
				await lib();

				expect(mockExitWithSuccess).toHaveBeenCalled();
			});

			it("prints pinned dependencies if --peer-deps is present", async () => {
				doMockCommands({
					"peer-deps": true,
				});

				await lib();

				expect(mockExitWithError).toHaveBeenCalled();
			});
		});

		describe("optional dependencies", () => {
			beforeEach(() => {
				doMockGetPackages({
					packages: [
						{
							dir: "/path/to/dir",
							relativeDir: "dir",
							packageJson: {
								name: "test",
								version: "1.0.0",
								optionalDependencies: {
									lodash: "^4.0.0",
								},
							},
						},
					],
				});
			});

			it("does not print pinned dependencies by default", async () => {
				await lib();

				expect(mockExitWithSuccess).toHaveBeenCalled();
			});

			it("prints pinned dependencies if --optionalDeps=true", async () => {
				doMockCommands({
					"optional-deps": true,
				});

				await lib();

				expect(mockExitWithError).toHaveBeenCalled();
			});
		});
	});

	describe("monorepo", () => {
		it("includes the root package.json if found", async () => {
			doMockGetPackages({
				rootPackage: {
					dir: "/path/to/root",
					relativeDir: "root",
					packageJson: {
						name: "root",
						version: "1.0.0",
						dependencies: {
							"root-dep": "^1.0.0",
						},
					},
				},
				packages: [
					{
						dir: "/path/to/dir",
						relativeDir: "dir",
						packageJson: {
							name: "test",
							version: "1.0.0",
							dependencies: {
								"test-dep": "^1.0.0",
							},
						},
					},
				],
			});

			await lib();

			expect(mockLog.mock.calls.flat()).toMatchInlineSnapshot(`
				[
				  "Package: /path/to/root/package.json",
				  "→ root-dep@^1.0.0",
				  "Package: /path/to/dir/package.json",
				  "→ test-dep@^1.0.0",
				]
			`);
		});
	});
});

function doMockCommands(commands: Partial<CliConfigType> = {}) {
	const defaultConfig = {
		"peer-deps": false,
		"no-deps": false,
		"no-dev-deps": false,
		"optional-deps": false,
	} as CliConfigType;

	const config = {
		...defaultConfig,
		...commands,
	};

	mockGetCliConfig.mockReturnValue(config);
}

function doMockGetPackages(mockReturn: Partial<GetPackages> = {}) {
	const defaultReturn = {
		packages: [],
		rootPackage: undefined,
	} satisfies Partial<GetPackages>;

	mockGetPackages.mockReturnValue({
		...defaultReturn,
		...mockReturn,
	});
}
