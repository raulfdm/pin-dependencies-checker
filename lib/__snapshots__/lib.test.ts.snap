// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`lib no pinned lib prints a nice message 1`] = `
[
  "All dependencies are pinned! ğŸ™Œ",
]
`;

exports[`lib found pinned lib prod dependencies prints all pinned dependencies by default 1`] = `
[
  "ğŸ‘® It seems you have unpinned dependencies. Please remove the caret from then.",
  
"-------- File --------
/path/to/dir/package.json

-------- Dependencies --------
"test-dep": "^1.0.0""
,
]
`;

exports[`lib found pinned lib dev dependencies prints all pinned dependencies by default 1`] = `
[
  "ğŸ‘® It seems you have unpinned dependencies. Please remove the caret from then.",
  
"-------- File --------
/path/to/dir/package.json

-------- Dependencies --------
"test-dev-dep": "^2.0.0"
"another-test-dev-dep": "^1.0.0""
,
]
`;
